pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/ExcessivePublicCountRule.java

```diff
@@ -25,44 +30,30 @@
  *
  * @author aglover
  */
-public class ExcessivePublicCountRule extends ExcessiveNodeCountRule {
+public class ExcessivePublicCountRule extends AbstractJavaCounterCheckRule<ASTAnyTypeDeclaration> {
 
     public ExcessivePublicCountRule() {
-        super(ASTCompilationUnit.class);
-        setProperty(MINIMUM_DESCRIPTOR, 45d);
+        super(ASTAnyTypeDeclaration.class,
+              ASTEnumDeclaration.class,
+              ASTClassOrInterfaceDeclaration.class,
+              ASTAnnotationTypeDeclaration.class);
     }
 
-    /**
-     * Method counts ONLY public methods.
-     */
     @Override
-    public Object visit(ASTMethodDeclarator node, Object data) {
-        return this.getTallyOnAccessType((AccessNode) node.jjtGetParent());
+    protected int defaultReportLevel() {
+        return 45;
     }
 
-    /**
-     * Method counts ONLY public class attributes which are not PUBLIC and
-     * static- these usually represent constants....
-     */
     @Override
-    public Object visit(ASTFieldDeclaration node, Object data) {
-        if (node.isFinal() && node.isStatic()) {
-            return NumericConstants.ZERO;
-        }
-        return this.getTallyOnAccessType(node);
-    }
+    protected boolean isViolation(ASTAnyTypeDeclaration node, int reportLevel) {
+        long publicCount = node.getDeclarations().stream()
+                               .map(ASTAnyTypeBodyDeclaration::getDeclarationNode)
+                               .filter(Objects::nonNull)
+                               .filter(it -> it instanceof AccessNode && ((AccessNode) it).isPublic())
+                               // filter out constants
+                               .filter(it -> !(it instanceof ASTFieldDeclaration && ((ASTFieldDeclaration) it).isFinal() && ((ASTFieldDeclaration) it).isStatic()))
+                               .count();
 
-    /**
-     * Method counts a node if it is public
-     *
-     * @param node
-     *            The access node.
-     * @return Integer 1 if node is public 0 otherwise
-     */
-    private Integer getTallyOnAccessType(AccessNode node) {
-        if (node.isPublic()) {
-            return NumericConstants.ONE;
-        }
-        return NumericConstants.ZERO;
+        return publicCount >= reportLevel;
```

Comment:
This piece depends on #1622, which is currently on hold as we have issues with streams and Java 8 & 9. Otherwise this PR should be ready to merge. Consider rewriting it so we may proceed.

Reply:
This doesn't depend on #1622, it uses regular java streams. Also the issues you're talking about only affect `Stream::flatMap`, which is not used by this piece of code. 

Besides, the flatmap problem is only a problem if either (1) your stream is infinite, or (2) you're performing side-effects while processing. (1) is never true for node streams and (2) is rarely the case. I don't think stalling #1622 just because of that is sensible. In any case we could override `NodeStream::flatMap` with the corrected implementation (just like what's done for `takeWhile`).

Reply:
my bad, you are right regarding this not depending on #1622. We should discuss #1622 in the respective PR.